acme4j-utils/src/test/java/org/shredzone/acme4j/util/CSRBuilderTest.java

```diff
@@ -111,6 +91,115 @@ public void testGenerate() throws IOException {
      */
     @Test
     public void testECCGenerate() throws IOException {
+        CSRBuilder builder = createBuilderWithValues();
+
+        builder.sign(testEcKey);
+
+        PKCS10CertificationRequest csr = builder.getCSR();
+        assertThat(csr).isNotNull();
+        assertThat(csr.getEncoded()).isEqualTo(builder.getEncoded());
+
+        csrTest(csr);
+        writerTest(builder);
+    }
+
+    /**
+     * Make sure an exception is thrown when no domain is set.
+     */
+    @Test
+    public void testNoDomain() throws IOException {
+        IllegalStateException ise = assertThrows(IllegalStateException.class, () -> {
+            CSRBuilder builder = new CSRBuilder();
+            builder.sign(testKey);
+        });
+        assertThat(ise.getMessage())
+            .isEqualTo("No domain or IP address was set");
+    }
+
+    /**
+     * Make sure an exception is thrown when an unknown identifier type is used.
+     */
+    @Test
+    public void testUnknownType() {
+        IllegalArgumentException iae = assertThrows(IllegalArgumentException.class, () -> {
+            CSRBuilder builder = new CSRBuilder();
+            builder.addIdentifier(new Identifier("UnKnOwN", "123"));
+        });
+        assertThat(iae.getMessage())
+            .isEqualTo("Unknown identifier type: UnKnOwN");
+    }
+
+    /**
+     * Make sure all getters will fail if the CSR is not signed.
+     */
+    @Test
+    public void testNoSign() throws IOException {
+        CSRBuilder builder = new CSRBuilder();
+        IllegalStateException ise;
+
+        assertThatExceptionOfType(IllegalStateException.class)
+            .isThrownBy(builder::getCSR)
+            .as("getCSR()")
+            .withMessage("sign CSR first");
+        
+        assertThatExceptionOfType(IllegalStateException.class)
+            .isThrownBy(builder::getEncoded)
+            .as("getCSR()")
+            .withMessage("sign CSR first");
+
+        ise = assertThrows(IllegalStateException.class, () -> {
+            try (StringWriter w = new StringWriter()) {
+                builder.write(w);
+            }
+        }, "write()");
+        assertEquals("unexpected exception message", "sign CSR first", ise.getMessage());
```

Comment:
This one is still missing. :wink: The problem is that the project won't even compile as long as there is a `org.junit.Assert.assertEquals` import.